File: sdk/python/dev/src/python_sdk_dev/docs.py
from typing import Annotated

import dagger
from dagger import Doc, function, object_type

from .utils import mounted_workdir


@object_type
class Docs:
    """Manage the reference documentation (Sphinx)."""

    container: dagger.Container

    @function
    def build(self) -> dagger.Directory:
        """Build the documentation."""
        return (
            self.container.with_workdir("docs")
            .with_exec(["uv", "run", "sphinx-build", "-v", ".", "/dist"])
            .directory("/dist")
        )

    @function
    def preview(
        self,
        bind: Annotated[
            int,
            Doc("The port to bind the web preview for the built docs"),
        ] = 8000,
    ) -> dagger.Service:
        """Build and preview the documentation in the browser."""
        return (
            self.container.with_(mounted_workdir(self.build()))
            .with_exec(["uv", "run", "python", "-m", "http.server", str(bind)])
            .with_exposed_port(bind)
            .as_service()
        )

File: .dagger/modules/python-sdk-dev/docs.go
package main

import (
	"fmt"

	"dagger/python-sdk-dev/internal/dagger"
)

type Docs struct {
	// +private
	Container *dagger.Container
}

// Build the documentation
func (d *Docs) Build() *dagger.Directory {
	return d.Container.
		WithWorkdir("docs").
		WithExec(uvRun(
			"sphinx-build", "-v", ".", "/dist",
		)).
		Directory("/dist")
}

// Build and preview the documentation in the browser
func (d *Docs) Preview(
	// The port to bind the web preview for the built docs
	// +default=8000
	bind int,
) *dagger.Service {
	return d.Container.
		With(mountedWorkdir(d.Build())).
		WithExposedPort(bind).
		AsService(dagger.ContainerAsServiceOpts{Args: uvRun(
			"python", "-m", "http.server", fmt.Sprintf("%d", bind),
		)})
}
